library/bn_mul.h

```diff
@@ -666,55 +675,63 @@
  * x !=0, which we can detect using __OPTIMIZE__ (which is also defined by
  * clang and armcc5 under the same conditions).
  *
- * So, only use the optimized assembly below for optimized build, which avoids
- * the build error and is pretty reasonable anyway.
+ * If gcc needs to use r7, we use r1 as a scratch register and have a few extra
+ * instructions to preserve/restore it; otherwise, we can use r7 and avoid
+ * the preserve/restore overhead.
  */
-#if defined(__GNUC__) && !defined(__OPTIMIZE__)
-#define MULADDC_CANNOT_USE_R7
-#endif
-
-#if defined(__arm__) && !defined(MULADDC_CANNOT_USE_R7)
-
-#if defined(__thumb__) && !defined(__thumb2__)
+#define MULADDC_SCRATCH         "RS .req r1         \n\t"
```

Comment:
It's limited to the surrounding `asm(...)` block. I've tested the following on clang and gcc, which is a compile fail because the second asm block doesn't know about `RS`.

```
    asm volatile ("RS .req x1 \n\t"
                  "neg RS, x0 \n\t");
    asm volatile ("neg RS, x0 \n\t");
```